### PR TITLE
🎛️ RFC 78 - Initial Stimulus adoption v2

### PR DESCRIPTION
**Pull request by lb-**
## [RFC 78](https://github.com/wagtail/rfcs/pull/78) - Initial Stimulus adoption

This PR is a draft for the initial first set of commits to get Stimulus started in Wagtail.

## Overview

* This contains one Stimulus controller for the auto-submit form behaviour in images/documents listing drop-downs (e.g. changing the collection will submit the form) and is intentionally a simple case to get initial feedback
* This is a partial re-implementation of https://github.com/lb-/wagtail/pull/5 - a more fleshed out adoption with more code usage (header search, skip link, upgrade notification).
* This is also a smaller scoped version of https://github.com/wagtail/wagtail/pull/9075 (no docs, and no 'magic' controller inclusions).
* This intentionally includes no documentation for now except for a folder README.md (planned new convention discussed within the UI team) to help guide contributors.
* There are three non-core commits, two of them ignore some linting so that we can adopt Stimulus without more global ignoring of rules we should ignore, it also contains a fix for the draftail types that was causing CI failures previously.
* Builds on https://github.com/wagtail/wagtail/pull/9482 & https://github.com/wagtail/wagtail/pull/9483

### How to evaluate this PR

1. If you have not already, read the RFC https://github.com/wagtail/rfcs/pull/78
2. Read the Stimulus docs (or at least the first few pages of the 'handbook' section) https://stimulus.hotwired.dev/handbook/origin
3. Look at the granular commits, each builds on each other to adopt Stimulus in the code
4. Run the code locally `npm install` `npm start` will be required and check the behaviour of the drop-downs in the images/documents listing actually works
5. Read through the documentation and see if you can add your own controller easily within the bakerydemo
6. If the feedback is general - best to put it on the RFC - otherwise feel free to put implementation specific feedback on this PR and it will be incorporated back into the RFC in due course.

### CODE REVIEW

**Comment by lb-** [comment]
With RFC 78 merged - will mark this as ready for review.

**Comment by lb-** [comment]
Stimulus 3.1.1 out. Should update - https://github.com/hotwired/stimulus/releases/tag/v3.1.1

**Comment by lb-** [comment]
Rebased & updated to Stimulus 3.1.1

**Comment by lb-** [comment]
Rebased & updated to Stimulus 3.2

**Comment by lb-** [comment]
CI failure is unrelated to this change.

```
Starting postgres service container
  /usr/bin/docker pull postgres:11
  Error response from daemon: Head "https://registry-1.docker.io/v2/library/postgres/manifests/11": received unexpected HTTP status: 503 Service Unavailable
  Warning: Docker pull failed with exit code 1, back off 2.89 seconds before retry.
  /usr/bin/docker pull postgres:11
  Error response from daemon: Head "https://registry-1.docker.io/v2/library/postgres/manifests/11": received unexpected HTTP status: 503 Service Unavailable
  Warning: Docker pull failed with exit code 1, back off 2.837 seconds before retry.
  /usr/bin/docker pull postgres:11
  Error response from daemon: Head "https://registry-1.docker.io/v2/library/postgres/manifests/11": received unexpected HTTP status: 503 Service Unavailable
  Error: Docker pull failed with exit code 1
```

**Comment by pySilver** [comment]
Just a side note. I'm working with Django & Stimulus (well it's not binded in any way) and I'm so so so so happy with it. HTML-over-the-wire it is so simple and flexible approach. I'm very happy it's going to be used inside Wagtail (one of the best CMS/F on the market). It would be much easier to extend Admin panel having Stimulus integrated.

**Comment by lb-** [comment]
Thanks @pySilver  - very glad to hear that - if you have not already, be sure to have a read of the early draft documentation in the RFC.

Here is the direct link https://github.com/wagtail/rfcs/blob/main/text/078-adopt-stimulus-js.md#a-documentation-for-developers

The target audience of that section is those using Wagtail and wanting to customise parts of the admin and/or use the Stimulus application instance provided for their own JS widgets. With or without a Node tooling setup.

**Comment by lb-** [comment]
Rebased on latest (a few items have been merged in) thanks Thibuad.

File: client/src/includes/stimulus.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  document.dispatchEvent(new CustomEvent('wagtail:stimulus-init'));
+
+  /**
+   * Add event listener to allow any custom code to trigger debug mode easily.
+   */
+  window.addEventListener('wagtail:stimulus-enable-debug', () => {
+    // eslint-disable-next-line no-param-reassign
+    application.debug = true;
+    application.logDebugActivity('debug enabled', 'logDebugActivity');
+  });
```
Comment (thibaudcolas) [inline client/src/includes/stimulus.ts:126]: This strikes me as a good example of where exposing the application on the window object would be much simpler. "`Stimulus.debug = true` in the console" is what the Stimulus docs explicitly recommend.

File: client/src/includes/stimulus.test.js [hunk 2]
```diff
@@ -0,0 +1,253 @@
+import { Application, Controller } from '@hotwired/stimulus';
```
Comment (thibaudcolas) [inline client/src/includes/stimulus.test.js:1]: Is there a reason why we’re not writing those tests in TypeScript if the corresponding new code is written in TypeScript?

File: client/src/includes/stimulus.test.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  const stimulusInit = jest.fn();
+  document.addEventListener('wagtail:stimulus-init', stimulusInit);
+
+  const stimulusReady = jest.fn(({ detail }) => {
+    // register example controllers for other tests
+    detail.register('example-a', detail.createController(wordCountController));
+    detail.register('example-b', WordCountController);
+  });
+
+  document.addEventListener('wagtail:stimulus-ready', stimulusReady);
```
Comment (thibaudcolas) [inline client/src/includes/stimulus.test.js:108]: Here and above – I think this code should be in a `beforeEach` or `beforeAll`? Jest can run tests in parallel, I’m not sure what happens when using globals like `document.body` at the top level like this.

**Review by thibaudcolas — COMMENTED** [review]
All working exactly as intended as far as I can tell! There are a few things I’d like to suggest simplifications for, keen to hear what you think.

File: .eslintrc.js [hunk 4]
```diff
@@ -99,6 +99,41 @@ module.exports = {
...
     },
+    // Rules we want to enforce or change for Stimulus Controllers
+    {
+      files: ['*Controller.ts'],
+      rules: {
+        '@typescript-eslint/member-ordering': [
+          'error',
+          {
+            classes: {
+              memberTypes: ['constructor', 'field', 'method', 'signature'],
```
[the remark this answers is not in the document]
Comment (lb-) [reply .eslintrc.js:110]: good point - I mainly just copied this from the docs for this (I thought some order is better than no order).

I think there could be small edge cases where we may still want constructor and if we pull this eslint rule up to all classes in code it would be nice to have the same approach.

File: client/src/includes/stimulus.test.js (hunk 2 above)
Comment (lb-) [reply client/src/includes/stimulus.test.js:1]: As per comment on other PR (cannot remember which one) - I understood we write unit tests in JS - we only have a very small subset of tests written in TS (2) vs most of them in JS (50+).

I also think it's easier (developer overhead) to write these quickly in JS and allows you to test things that maybe should not happen in theory in TS but could happen when called externally.

My preference would be for us to be clear that we do want all tests in TS before we start doing this - maybe worth discussing when we work on the UI development guide.

https://docs.google.com/document/d/1Sujt_wgeRvTWMFH4v_Br6z7IpEBRdHq-y-OyOpL-Y6o/edit#

File: client/src/includes/stimulus.test.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    this.element.insertAdjacentElement('beforebegin', output);
+    this.output = output;
+  },
+  updateCount(event) {
+    const value = event ? event.target.value : this.element.value;
+    const words = (value || '').split(' ');
+    this.output.textContent = `${words.length} / ${this.maxValue} words`;
+  },
+  disconnect() {
+    this.output && this.output.remove();
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.test.js:31]: It should not happen, however, in theory something else could remove this element from the DOM before the disconnect method is called.

File: client/src/controllers/AutoFieldController.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * @example
+ * // once any change is made to the below select field, the form will be auto submitted
+ * <form>
+ *   <select name="order" data-controller="w-auto-field" data-action="change->w-auto-field#submit">
+ *     <option value="A-Z">A to Z</option>
+ *     <option value="Z-A">Z to A</option>
+ *   </select>
+ * </form>
+ */
+export class AutoFieldController extends Controller<HTMLInputElement> {
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/AutoFieldController.ts:15]: This may not technically be correct - the current use case is only for select elements but nothing is stopping us from using this on any input.

File: client/src/controllers/AutoFieldController.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  submit() {
+    this.element.form.submit();
```
Comment (lb-) [inline client/src/controllers/AutoFieldController.ts:17]: ok this breaks change breaks the types - I will have to rework this

File: client/src/includes/stimulus.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+ * Turns on debug mode if in local development (for now).
+ */
+export const initStimulus = ({
+  debug = process.env.NODE_ENV === 'development',
+  definitions = [],
+  element = document.documentElement,
+} = {}) => {
+  const application = Application.start(element);
+
+  application.debug = !!debug;
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts:194]: Good call thanks

File: client/src/controllers/AutoFieldController.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply client/src/controllers/AutoFieldController.ts:15]: Perhaps `HTMLInputElement | HTMLSelectElement` then?

File: client/src/includes/stimulus.test.js (hunk 2 above)
Comment (lb-) [reply client/src/includes/stimulus.test.js:1]: As per chat - you are happy with this for now @thibaudcolas

File: client/src/includes/stimulus.test.js (hunk 3 above)
Comment (lb-) [reply client/src/includes/stimulus.test.js:108]: Fixed

File: client/src/controllers/AutoFieldController.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/AutoFieldController.ts:15]: Great idea.

File: client/src/entrypoints/admin/core.js [hunk 9]
```diff
@@ -1,9 +1,15 @@
...
+
+import { coreControllerDefinitions } from '../../controllers';
 import { escapeHtml } from '../../utils/text';
 import { initButtonSelects } from '../../includes/initButtonSelects';
+import { initStimulus } from '../../includes/stimulus';
 import { initTagField } from '../../includes/initTagField';
 import { initTooltips } from '../../includes/initTooltips';
 
+/** initialise Wagtail Stimulus application with core controller definitions */
+initStimulus({ definitions: coreControllerDefinitions });
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/entrypoints/admin/core.js:11]: As `core.js` is loaded first I think it would be best that this be included there for now, is it OK to keep it in `core.js`? Assuming we get good progress with cleaning up the core.js my guess is that we would merge the files into one eventually anyway.

As for the DOMContentLoaded - no need - that all happens for us with Stimulus - you can init the application as early or as late as you want it will only start attaching to the DOM when it's ready.

It's actually a really simple Promise approach - see `domReady` and where it's used https://github.com/hotwired/stimulus/blob/main/src/core/application.ts#L110


---

wagtail/admin/templates/wagtailadmin/admin_base.html

```
    <script src="{% versioned_static 'wagtailadmin/js/core.js' %}"></script>
    <script src="{% versioned_static 'wagtailadmin/js/vendor.js' %}"></script>
    <script src="{% versioned_static 'wagtailadmin/js/wagtailadmin.js' %}"></script>
```

File: client/src/includes/stimulus.stories.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+          minWidth: '4ch',
+          textAlign: 'center',
+        }}
+      />
+    </p>
+  </StimulusWrapper>
+);
+
+export default {
+  title: 'Stimulus/Example',
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js:138]: Oh but this one is so fun :D 🎲

File: client/src/includes/stimulus.stories.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    // convert convenience object into identifier / controllerConstructor array if needed
+    const definitions = Array.isArray(controllers)
+      ? controllers
+      : Object.entries(controllers).map(
+          ([identifier, controllerConstructor]) => ({
+            identifier,
+            controllerConstructor,
+          }),
+        );
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js:38]: haha yeah - I actually wrote it that way initially but thought it would be too verbose for those working with this code and opted for the simpler syntax.

I will fix this up tomorrow.

File: client/src/includes/stimulus.stories.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+   */
+  controllers: PropTypes.oneOfType([
+    PropTypes.objectOf(PropTypes.func),
+    PropTypes.arrayOf(
+      PropTypes.shape({
+        controllerConstructor: PropTypes.func,
+        identifier: PropTypes.string,
+      }),
+    ),
+  ]),
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js:77]: Good call, I am used to being loose with the prop types in my work project but will sharpen this up tomorrow.

File: client/src/includes/stimulus.stories.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+ * const Template = () => <Stories.StimulusWrapper controllers={{ 'w-auto-field': AutoFieldController }}><form data-controller="w-auto-field" /></Stories.StimulusWrapper>
+ */
+class StimulusWrapper extends React.Component {
+  constructor(props) {
+    super(props);
+    this.ref = React.createRef();
+  }
+
+  componentDidMount() {
+    const { controllers = [], debug = false } = this.props;
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js:28]: yeah, I did that initially but must have turned it off, will fix up.

File: client/src/includes/stimulus.stories.js [hunk 14]
```diff
@@ -0,0 +1,153 @@
+/**
+ * Example Stimulus usage within Storybook, including StimulusWrapper export
+ * Note: Must use .js not .tsx for now
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js:3]: this was because of the ES6/ES6 disconnect - we may be able to use tsx now - I will revisit tomorrow.

File: client/src/includes/stimulus.ts (hunk 1 above)
Comment (lb-) [reply client/src/includes/stimulus.ts:126]: OK - on the `window.Stimulus` discussion - I think I need to heed the advice here and give this a go. I just want to double check - we are ok with adding yet another global in our quest to reduce globals across Wagtail. This new global being the exception :) 

If we are, can we keep this focused and attach the other 'things' we need to the instance (or as a static method).

We would provide the application instance as `window.Stimulus` and also provide the following;

* `window.Stimulus.createController` - the helper function that creates a controller from a plain object for those that do not want to think about classes or build tooling.
* `window.Stimulus.Controller` - this would be made available so that those who want to create a class can easily extend this with their own class (note: this will break for any Node build system based code that is NOT using ES6 or above as their compile target - we can document this of course).

If we want 'control' over behaviour for things like `register` we can override that class method and inject custom behaviour easily in the future.

We lose the ability to be strict about what 'comes in' to our Stimulus instance (there are always way that devs can do this but its harder if its not at the global). They will be able to `stop` the application instance which would break all widgets globally, unload any controllers, use the `getControllerForElementAndIdentifier` to directly access controller instances more easily (this seems like it's ok but can lead to more code being tied to the controller name).

These may be low risk (if developers break their own code, that is up to them), but it is a risk nonetheless.

If projects want to use their own isolated Stimulus instance, they of course will need to use a different global name.

It will be harder to change/rename this global once it's there.

An alternative approach is to expose a faux Stimulus instance which only exposes the things we are ready to support as a formal API (even if undocumented).

File: client/src/includes/stimulus.ts [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  Object.entries(controllerDefinition)
+    .filter(([key]) => !['STATIC'].includes(key))
+    .map(([key]) => ({
+      key,
+      property:
+        Object.getOwnPropertyDescriptor(controllerDefinition, key) || noop,
+    }))
+    .forEach(({ key, property }) => {
+      Object.defineProperty(NewController.prototype, key, property);
+    });
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts:51]: great call - a little bit of a rework required but this really helped & makes the code simpler.

File: client/src/includes/stimulus.ts (hunk 1 above)
Comment (lb-) [reply client/src/includes/stimulus.ts:126]: This will be updated soon so I will resolve this issue.

API will be via the `constructor` so;

`window.Stimulus.constructor.Controller`  (base class to extend)
`window.Stimulus.constructor.createController` (from object)

so full usage would be something like

```js
// create a controller from object and register it
(() => {
  const WordCountController = window.Stimulus.constructor.createController({
    STATIC: {
      values: { max: { default: 10, type: Number } },
    },
    connect() {
      console.log(`MaxValue: ${this.maxValue} - connected to:`, this.element);
    },
  });
  window.Stimulus.register('example-a', WordCountController);
})();
```

OR

```js
// create a controller class and register it (will not work if using Node tooling that compiles down to ES5)
(() => {
  class WordCountController extends window.stimulus.constructor.Controller {
    static values = { max: { default: 10, type: Number } };
    connect() {
      console.log(`MaxValue: ${this.maxValue} - connected to:`, this.element);
    }
  }
  window.Stimulus.register('example-a', WordCountController);
})();
```

File: client/src/includes/stimulus.stories.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js]: Ok. Turns out we do need .js still - it's an incompatibility between storybook's TS compiling and the core TS compiling. 

I have kept these as JS but pulled out the wrapper to TSX as that works fine.

File: client/src/includes/stimulus.stories.js [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+      defaultValue: true,
+    },
+    number: {
+      control: { type: 'select' },
+      description: 'Dice sides',
+      options: [2, 4, 6, 10, 20],
+    },
+  },
+  StimulusWrapper, // available for other stories to import
+};
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js]: I've pulled this out to a seperate file. Makes more sense this way. Thanks for the feedback.

File: client/src/controllers/AutoFieldController.stories.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    },
+  },
+};
+
+const Template = ({ debug = false }) => {
+  const [submitCount, updateSubmitCount] = useState(0);
+
+  return (
+    <Stories.StimulusWrapper
+      controllers={{ 'w-auto-field': AutoFieldController }}
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/AutoFieldController.stories.js]: Good call. This is now done and typed all the way through much better. Autocomplete in the editor yay.

File: client/src/includes/stimulus.stories.js [hunk 18]
```diff
@@ -0,0 +1,153 @@
+/**
+ * Example Stimulus usage within Storybook, including StimulusWrapper export
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.stories.js]: This is done.

File: client/src/includes/stimulus.ts [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+  ) => {
+    if (!controller || !identifier) {
+      application.logDebugActivity(
+        'registration failed',
+        controller ? 'no identifier' : 'no controller',
+        { controller, identifier },
+      );
+      return;
+    }
+    application.load({ identifier, controllerConstructor: controller });
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts]: Load is better when dealing with a potential collection of controllers pulled in from other code. 

Register is good for one at a time. 

See https://stimulus.hotwired.dev/handbook/installing#using-webpack-helpers (note the use of load there).

File: client/src/includes/stimulus.ts [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    .forEach(({ key, property }) => {
+      Object.defineProperty(NewController.prototype, key, property);
+    });
+
+  return NewController;
+};
+
+/**
+ * Dispatches events and adds event listeners for documented API surface without relying
+ * on exposing the application instance or adding more globals.
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts]: As per other comments. This is now done.

File: client/src/includes/stimulus.ts [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+    application.logDebugActivity('registered', identifier);
+  };
+
+  /**
+   * Dispatch event for stimulus being initialised.
+   * Most code will not be able to listen to this event,
+   * setting up for any future fallbacks/provision of application or
+   * cancelling/modifying core controllers.
+   */
+  document.dispatchEvent(new CustomEvent('wagtail:stimulus-init'));
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts]: These events are now all removed.

File: client/src/includes/stimulus.ts [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+ * on exposing the application instance or adding more globals.
+ *
+ * @example - within Wagtail for custom stimulus controllers
+ * ```python
+ * @hooks.register('insert_global_admin_js')
+ * def global_admin_js():
+ *     return mark_safe(
+ *         """
+ * <script type="module">
+ * import { Controller } from "https://unpkg.com/@hotwired/stimulus/dist/stimulus.js";
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/includes/stimulus.ts]: I still need to update these examples.

File: client/src/controllers/README.md [hunk 23]
```diff
@@ -0,0 +1,8 @@
+# `src/controllers` folder
+
+**Important:** This is a migration in progress, any large refactors or new code should adopt this approach.
+
+-   Wagtail uses [Stimulus](https://stimulus.hotwired.dev/) as a way to attach interactive behaviour to DOM elements.
+-   This is a lightweight JavaScript framework that allows a JavaScript class to be attached to any DOM element that adheres to a specific usage of `data-` attributes on the element.
+-   Each file within this folder should contain one Stimulus controller class, using a matching file name (for example `class MyAwesomeController, `MyAwesomeController.ts`, all TitleCase).
+-   Controllers that are included in the `index.ts` default export will automatically be included in the core bundle and provided by default.
```
Comment (lb-) [inline client/src/controllers/README.md:8]:
```suggestion
-   Controllers that are included in the `index.ts` default export will automatically be included in the core bundle and provided by default.
-  Stories need to be written as JavaScript for now - `MyController.stories.js` as the compiled JavaScript from StoryBook conflicts with Stimulus' usage of adding getters only on Controller instantiation.
```

**Review by lb- — COMMENTED** [review]
@thibaudcolas - this is ready for another review. 

I've added one comment for myself, the JSDoc is alll cleaned up also and I've intentionally avoided an example of createController / Wagtail hook usage for now

File: client/src/controllers/AutoFieldController.ts [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+export class AutoFieldController extends Controller<
+  HTMLInputElement | HTMLSelectElement
+> {
+  submit() {
+    if (!this.element.form) {
+      throw new Error(
+        `${this.identifier} controlled element must be part of a <form />`,
+      );
+    }
+    this.element.form.submit();
```
Comment (lb-) [inline client/src/controllers/AutoFieldController.ts:24]: @thibaudcolas  after a bit of thinking I will put back the `requestSubmit` thing - this does a nicer job of calling the submit event (for event handlers) and client-side validation can be avoided by `novalidate`

File: client/src/controllers/README.md (hunk 23 above)
Comment (lb-) [reply client/src/controllers/README.md:8]: This has been added.

File: client/src/controllers/AutoFieldController.ts (hunk 24 above)
Comment (lb-) [reply client/src/controllers/AutoFieldController.ts]: The usage of `requestSubmit` has been added back in.

**Review by thibaudcolas — APPROVED** [review]
I’m surprised you removed _all_ of the initialisation events – aside from that all good!

Note while testing this I got a PropTypes error from Storybook saying the React `checked` prop was submitted as the string "true" rather than a boolean. As far as I can tell this is a Storybook bug but thought I’d mention just in case.

Tested in Chrome 108 macOS 13, Firefox 102 Windows 11, Edge 102 Windows 11, Safari 14.1 macOS 12.1.

`source-map-explorer` output with Stimulus added for reference (nothing outstanding): https://leafy-centaur-e7a959.netlify.app/

**Comment by laymonage** [comment]
Aaand it's in! Thank you very much @lb- and @thibaudcolas, can't wait to use this for new code and some refactorings 🥳

**Comment by lb-** [comment]
@thibaudcolas thank you for the thoughtful feedback on this, it's great to see it go in. 

As for the events - I figured they may become helpful in the future, if so we can add them if needed. I've got a backup of the branch before removing them.
